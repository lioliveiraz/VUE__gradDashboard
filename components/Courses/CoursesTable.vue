<template>
  <table class="g-table-wrapper" data-testId="course_table">
    <thead>
      <tr class="font-bold">
        <th class="g-table-cell">
          {{ $t("TABLE_HEAD_WEEK") }}
        </th>
        <th class="g-table-cell">
          {{ $t("TABLE_HEAD_COURSECODE") }}
        </th>
        <th class="g-table-cell">
          {{ $t("TABLE_HEAD_COURSENAME") }}
        </th>
        <th class="g-table-cell">
          {{ $t("TABLE_HEAD_MODE") }}
        </th>
        <th class="g-table-cell">
          {{ $t("TABLE_HEAD_SOURCE") }}
        </th>
        <th class="g-table-cell">
          {{ $t("TABLE_HEAD_DURATION") }}
        </th>
        <th class="g-table-cell">
          {{ $t("TABLE_HEAD_LINK") }}
        </th>
      </tr>
    </thead>
    <tbody>
      <tr class="bg-emerald-200" v-for="course in courses" :key="course.id">
        <td class="g-table-cell" data-testId="course_table-week">
          {{ course.week }}
        </td>
        <td class="g-table-cell" data-testId="course_table-code">
          {{ course.course_code }}
        </td>
        <td class="g-table-cell" data-testId="course_table-name">
          {{ course.course_name }}
        </td>
        <td class="g-table-cell" data-testId="course_table-assessment">
          {{ course.assessment ? $t("ASSESSMENTS_VALUE") : "self-paced" }}
        </td>
        <td class="g-table-cell" data-testId="course_table-source">
          {{ course.source }}
        </td>
        <td class="g-table-cell" data-testId="course_table-duration">
          {{ course.duration }}h
        </td>
        <td class="g-table-cell" data-testId="course_table-link">
          <a :href="course.link">link</a>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
export default {
  props: { courses: Array },
};
</script>

<style>
@import "../../assets/styles/tables.css";
</style>