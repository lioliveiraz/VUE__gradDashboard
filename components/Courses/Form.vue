<template>
  <form @submit="handleSubmit" data-testId="course_form p-2" :key="key">
    <h3 class="text-green-900 font-bold text-2xl">
      Add new Course or Assessment
    </h3>
    <Input
      @getUserInput="getUserInput"
      :attributeObj="{
        type: this.NUMBER_INPUT,
        name: this.TABLE_HEAD_WEEK_ENGLISH,
        placeholder: 'week number',
        required: true,
      }"
    />
    {{ errors.week && errors.week }}

    <Input
      @getUserInput="getUserInput"
      :attributeObj="{
        type: this.TEXT_INPUT,
        name: this.COURSE_CODE_INPUT,
        required: true,
      }"
    />
    {{ errors.course_code && errors.course_code }}

    <Input
      @getUserInput="getUserInput"
      :attributeObj="{
        type: this.TEXT_INPUT,
        name: this.COURSE_NAME_INPUT,
        required: true,
      }"
    />
    {{ errors.course_name && errors.course_name }}

    <Input
      @getUserInput="getUserInput"
      :attributeObj="{
        type: this.TEXT_INPUT,
        name: this.TABLE_HEAD_SOURCE_ENGLISH,
        placeholder: 'Udemy',
        required: true,
      }"
    />
    {{ errors.source && errors.source }}

    <Input
      @getUserInput="getUserInput"
      :attributeObj="{
        type: this.NUMBER_INPUT,
        name: this.TABLE_HEAD_DURATION_ENGLISH,
        placeholder: '875463',
        required: true,
      }"
    />
    <Input
      @getUserInput="getUserInput"
      :attributeObj="{ type: this.CHECKBOX_INPUT, name: this.ASSESSMENT_INPUT }"
    />
    <div class="flex">
      <input :type="this.BUTTON_SUBMIT" class="text-gray-100 w-48 h-12 m-1" />
      <button @click="toggleComponent" class="text-gray-100 w-48 h-12 m-1">
        Close
      </button>
    </div>
  </form>
</template>

<script>
import Input from "../Input";
import { addCourse } from "../../api/requests/post";
import { mapGetters, mapActions } from "vuex";
import { courseValidation } from "../../helpers/validation";

export default {
  components: { Input },
  props: ["toggleComponent"],

  data() {
    return {
      courseData: {
        assessment: false,
        link:
          "https://giphy.com/gifs/makespace-cat-yoga-xUPGcyi4YxcZp8dWZq/tile",
      },
      errors: {},
      key: 0,
    };
  },
  computed: {
    ...mapGetters("auth", ["getToken"]),
  },
  methods: {
    ...mapActions("courses", ["fetchCourses", "handleAddCourse"]),

    getUserInput(inputValue, inputName) {
      this.courseData[inputName] = inputValue;
    },
    async handleSubmit(e) {
      e.preventDefault();
      this.courseData[this.DURATION_INPUT] = +this.courseData[
        this.DURATION_INPUT
      ];
      const errors = courseValidation(this.courseData);
      if (Object.entries(errors).length !== 0) {
        this.errors = errors;
      } else {
        addCourse(this.courseData, this.getToken)
          .then((res) => {
            this.errors = {};
            this.toggleComponent();
            this.key++;
            this.handleAddCourse(this.courseData);
            this.$toast(res.data.message, { type: this.TOAST_SUCCESS });
          })
          .catch((err) =>
            this.$toast(err.response.data.message, { type: this.TOAST_ERROR })
          );
      }
    },
  },
};
</script>

<style scoped>
input[type="submit"] {
  background: #00c300;
}
form {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
}
form h3 {
  text-align: center;
}
button {
  background: #00c300;
}
@media screen and (min-width: 700px) {
  form {
    width: 70%;
  }
}
</style>