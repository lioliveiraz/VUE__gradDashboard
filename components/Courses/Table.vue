<template>
  <table class="table-auto border-separate" data-testId="course_table">
    <thead>
      <tr>
        <th class="border border-teal-600 p-1 md:p-4">
          {{ this.TABLE_HEAD_WEEK_ENGLISH }}
        </th>
        <th class="border border-teal-600 p-1 md:p-4">
          {{ this.TABLE_HEAD_COURSECODE_ENGLISH }}
        </th>
        <th class="border border-teal-600 p-1 md:p-4">
          {{ this.TABLE_HEAD_COURSENAME_ENGLISH }}
        </th>
        <th class="border border-teal-600 p-1 md:p-4">
          {{ this.TABLE_HEAD_MODE_ENGLISH }}
        </th>
        <th class="border border-teal-600 p-1 md:p-4">
          {{ this.TABLE_HEAD_SOURCE_ENGLISH }}
        </th>
        <th class="border border-teal-600 p-1 md:p-4">
          {{ this.TABLE_HEAD_DURATION_ENGLISH }}
        </th>
        <th class="border border-teal-600 p-1 md:p-4">
          {{ this.TABLE_HEAD_LINK_ENGLISH }}
        </th>
      </tr>
    </thead>
    <tbody>
      <tr class="bg-emerald-200" v-for="course in courses" :key="course.id">
        <td
          class="border border-teal-600 p-1 md:p-4"
          data-testId="course_table-week"
        >
          {{ course.week }}
        </td>
        <td
          class="border border-teal-600 p-1 md:p-4"
          data-testId="course_table-code"
        >
          {{ course.course_code }}
        </td>
        <td
          class="border border-teal-600 p-1 md:p-4"
          data-testId="course_table-name"
        >
          {{ course.course_name }}
        </td>
        <td
          class="border border-teal-600 p-1 md:p-4"
          data-testId="course_table-assessment"
        >
          {{ course.assessment ? "assessment" : "self-paced" }}
        </td>
        <td
          class="border border-teal-600 p-1 md:p-4"
          data-testId="course_table-source"
        >
          {{ course.source }}
        </td>
        <td
          class="border border-teal-600 p-1 md:p-4"
          data-testId="course_table-duration"
        >
          {{ course.duration }}h
        </td>
        <td
          class="border border-teal-600 p-1 md:p-4"
          data-testId="course_table-link"
        >
          <a :href="course.link">link</a>
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script>
export default {
  props: ["courses"],
};
</script>

<style>
th {
  text-transform: uppercase;
}
</style>