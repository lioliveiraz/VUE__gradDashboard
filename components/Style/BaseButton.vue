<template>
  <button class="g-button-base" @click="handleClick">
    {{ value }}
  </button>
</template>

<script>
export default {
  props: {
    handleClick: { type: Function, required: false },
    value: String,
  },
};
</script>

<style>
</style>