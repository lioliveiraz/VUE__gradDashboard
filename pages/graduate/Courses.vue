<template>
  <div v-if="courses">
    <button @click="toggleCourses">
      {{ !isCourse ? "courses" : "assesmentes" }}
    </button>
    <div v-if="isCourse">
      <Table :courses="courses" />
    </div>
    <div v-if="!isCourse">
      <Table :courses="assessments" />
      <select
        name="code"
        id=""
        v-for="assesment in assessments"
        :key="assesment.id"
      >
        <option :value="assesment.course_code">
          name :{{ assesment.course_name }} || code:{{ assesment.course_code }}
        </option>
      </select>
    </div>
  </div>
</template>

<script>
import { getCourses } from "../../api/requests";
import Table from "../../components/Table";

export default {
  layout: "graduate",
  watchQuery: ["dashboard"],
  components: {
    Table,
  },
  data() {
    return {
      isCourse: true,
      courses: null,
      assessments: null,
    };
  },
  async created() {
    const token = await this.$store.state.auth.token;
    if (token) {
      try {
        const res = await getCourses(token);
        const coursesArr = res.filter((course) => course.assessment === false);
        const assesmentsArr = res.filter(
          (course) => course.assessment === true
        );
        this.courses = coursesArr;
        this.assessments = assesmentsArr;
      } catch (error) {
        console.log(error);
      }
    }
  },

  methods: {
    toggleCourses() {
      this.isCourse = !this.isCourse;
    },
  },
};
</script>

<style>
</style>