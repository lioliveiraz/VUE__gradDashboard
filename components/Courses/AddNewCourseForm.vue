<template>
  <div class="relative w-auto my-6 mx-auto h-auto">
    <div class="g-modal-form--inner focus:outline-none">
      <div class="border-blueGray-200 g-modal-form--header">
        <h3 data-testId="course_form-title">Update Path</h3>
      </div>
      <div class="g-modal-form--content">
        <form @submit="handleSubmit" data-testId="course_form" :key="key">
          <div>
            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.NUMBER_INPUT,
                name: this.TABLE_HEAD_WEEK_ENGLISH,
                placeholder: 'week number',
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.TEXT_INPUT,
                name: this.COURSE_CODE_INPUT,
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.TEXT_INPUT,
                name: this.COURSE_NAME_INPUT,
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.TEXT_INPUT,
                name: this.TABLE_HEAD_SOURCE_ENGLISH,
                placeholder: 'Udemy',
                required: true,
              }"
            />

            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.NUMBER_INPUT,
                name: this.TABLE_HEAD_DURATION_ENGLISH,
                placeholder: '875463',
                required: true,
              }"
            />
            <BaseInput
              @getUserInput="getUserInput"
              :attributeObj="{
                type: this.CHECKBOX_INPUT,
                name: this.ASSESSMENT_INPUT,
              }"
            />

            <div class="g-addNewcourse-buttons">
              <input
                type="submit"
                value="Send"
                :class="
                  isFormValid ? 'g-base-btn-blue' : 'g-base-btn-blue g-disabled'
                "
                :disabled="!isFormValid"
              />
              <BaseButton :handleClick="toggleComponent" value="close" />
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</template>

<script>
import BaseInput from "../BaseInput";
import { addCourse } from "../../api/requests/post";
import { mapGetters, mapActions } from "vuex";
import { courseValidation } from "../../helpers/validation";
import BaseButton from "../../components/Style/BaseButton";
import { isObjectEmpty, isObjectValuesEmpty } from "../../helpers/service";

export default {
  components: { BaseInput, BaseButton },
  props: { toggleComponent: Function },

  data() {
    return {
      courseData: {
        assessment: false,
        link:
          "https://giphy.com/gifs/makespace-cat-yoga-xUPGcyi4YxcZp8dWZq/tile",
      },
      key: 0,
      isFormValid: false,
    };
  },
  computed: {
    ...mapGetters("auth", ["getToken"]),
  },
  methods: {
    ...mapActions("courses", ["fetchCourses", "handleAddCourse"]),

    formValidation() {
      const errors = courseValidation(this.courseData);
      this.isFormValid = isObjectEmpty(errors);
    },
    getUserInput(inputValue, inputName) {
      this.courseData[inputName] = inputValue;
      if (
        !isObjectEmpty(this.courseData) &&
        isObjectValuesEmpty(this.courseData).length === 7
      ) {
        this.formValidation();
      } else {
        this.isFormValid = false;
      }
    },
    async handleSubmit(e) {
      e.preventDefault();
      this.courseData[this.DURATION_INPUT] = +this.courseData[
        this.DURATION_INPUT
      ];
      if (!this.isFormValid) return null;
      addCourse(this.courseData, this.getToken)
        .then((res) => {
          this.errors = {};
          this.toggleComponent();
          this.key = this.key + 1;
          this.handleAddCourse(this.courseData);
          this.$toast(res.data.message, { type: this.TOAST_SUCCESS });
        })
        .catch((err) =>
          this.$toast(err.response.data.message, { type: this.TOAST_ERROR })
        );
    },
  },
};
</script>

<style scoped>
</style>