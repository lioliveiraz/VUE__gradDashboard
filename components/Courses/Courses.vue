<template>
  <div class="g-courses-wrapper">
    <div class="g-tables">
      <BaseButton
        :handleClick="toggleCourses"
        :value="!isCourse ? courseValue : assessmentValue"
      />

      <div class="g-tables-inner">
        <h3 class="text-gray-600 self-auto">
          {{ isCourse ? courseValue : assessmentValue }}
        </h3>
        <CoursesTable v-if="isCourse" :courses="coursesArr" />

        <CoursesTable v-if="!isCourse" :courses="assessmentsArr" />
      </div>
    </div>
  </div>
</template>

<script>
import CoursesTable from "./CoursesTable";
import BaseButton from "../../components/Style/BaseButton";

export default {
  props: {
    coursesArr: Array,
    assessmentsArr: Array,
  },
  data() {
    return {
      isCourse: true,
    };
  },
  components: {
    CoursesTable,
    BaseButton,
  },
  computed: {
    courseValue() {
      return this.$t("COURSES_VALUE");
    },
    assessmentValue() {
      return this.$t("ASSESSMENTS_VALUE");
    },
  },

  methods: {
    toggleCourses() {
      this.isCourse = !this.isCourse;
    },
  },
};
</script>

<style scoped>
</style>