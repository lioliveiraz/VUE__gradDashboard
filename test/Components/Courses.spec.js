import Courses from '../../components/Courses/Courses.vue';
import { shallowMount } from '@vue/test-utils';
import BaseButton from '../../components/Style/BaseButton.vue';
import CoursesTable from '../../components/Courses/CoursesTable.vue';

describe('<Courses>', () => {
    let wrapper, coursesArr, assessmentsArr;
    coursesArr = [];
    assessmentsArr = [];
    beforeEach(async () => {

        wrapper = await shallowMount(Courses, {
            propsData: { coursesArr, assessmentsArr }, mocks: {
                $t: (msg) => msg
            }
        });
    });
    it('should render correctly', () => {
        const courseTable = wrapper.findComponent(CoursesTable);
        const tables = wrapper.findAllComponents(CoursesTable);
        const button = wrapper.find("button");

        const baseButton = wrapper.findComponent(BaseButton);

        let idsArr = [
            ".g-courses-wrapper", ".g-tables-inner"];
        idsArr.forEach(id => {
            expect(wrapper.find(id)).toBeTruthy();

        });


        expect(baseButton.exists()).toBeTruthy();
        expect(courseTable.exists()).toBeTruthy();
        expect(tables).toHaveLength(1);
        expect(wrapper).toMatchSnapshot();

    });
    it('function should be called', async () => {
        await wrapper.vm.toggleCourses();
        expect(wrapper.vm.isCourse).toBe(false);
    });
});
