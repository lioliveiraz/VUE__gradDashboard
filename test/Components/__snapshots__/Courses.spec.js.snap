// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Courses> should render correctly 1`] = `
VueWrapper {
  "_emitted": Object {},
  "_emittedByOrder": Array [],
  "isFunctionalComponent": undefined,
}
`;
